File: PythonSolutions/Neetcode150/slidingwindow_problems/121_best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        start, end = 0, 1

        maxProfit = 0

        while end < len(prices):
            # Check if previous day price is less than the future date price
            # This means we were able to buy low and sell high
            if prices[start] < prices[end]:
                # Calculate profit
                maxProfit = max(maxProfit, prices[end] - prices[start])
            else:
                # If the previous day price is greater than the future date price
                # Then this means we are not able to buy low and sell high

                # So we update the price to the future date price since we found a better buy price as future date is lower
                start = end
            
            # Move to the next day
            end += 1
        
        return maxProfit

File: TypescriptSolutions/Neetcode150/sliding_window_problems/121_best-time-to-buy-and-sell-stock.ts
const maxProfit = (prices: number[]): number => {
    const n = prices.length;
    let start = 0;
    let end = 1;

    let maxProfit = 0;

    while (end < n) {
        // We are able to buy the stock for profit
        if (prices[start] < prices[end]) {
            maxProfit = Math.max(prices[end] - prices[start], maxProfit);
        }
        // There will be no profit, cheaper price in the future so wait for that day
        else {
            start = end;
        }

        // Keep incrementing the day to check for all larger priced days
        end++;
    }

    return maxProfit;
};

const readableDynamicWindowMaxProfitSolution = (prices: number[]): number => {
    let start = 0;
    let maxProfit = 0;

    for (let end = 1; end < prices.length; end++) {
        if (prices[start] < prices[end]) {
            maxProfit = Math.max(prices[end] - prices[start], maxProfit);
        } else {
            // Found a smaller stock price in the future so close dynamic window
            start = end;
        }
    }

    return maxProfit;
};